// @event-nest/postgresql/src/lib/schema-configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SchemaConfiguration = void 0;
/**
 * A configuration class which holds the schema, aggregates table, and events table names for the PostgreSQL database.
 */
class SchemaConfiguration {
    constructor(_schema, _aggregatesTable, _eventsTable) {
        this._schema = _schema;
        this._aggregatesTable = _aggregatesTable;
        this._eventsTable = _eventsTable;
    }
    get aggregatesTable() {
        return this._aggregatesTable;
    }
    get eventsTable() {
        return this._eventsTable;
    }
    get schema() {
        return this._schema;
    }
    /**
     * The schema-aware aggregates table name (schema.aggregatesTable)
     */
    get schemaAwareAggregatesTable() {
        return this._schema + "." + this._aggregatesTable;
    }
    /**
     * The schema-aware events table name (schema.eventsTable)
     */
    get schemaAwareEventsTable() {
        return this._schema + "." + this._eventsTable;
    }
}
exports.SchemaConfiguration = SchemaConfiguration;
